### ISSUE TITLE
Fail on missing files when using `--files-from=`

### ISSUE THREAD

**Issue by static-moonlight** [rclone/rclone#7842]
When coping files with `rclone --files-from=` the [documentation](https://rclone.org/filtering/) states:

> Rclone commands do not error if any names in the --files-from file are missing from the source remote.

However, that is a problem for me. I need to make sure that all files have been copied. So this is a request to put in an option to let `--files-from=` fail in case files are missing.

I'm also open for any suggestions how to verify that all files (from the given list) are present at the target location.

**Comment by jeroendeswaef** [comment]
I'm experiencing the same issue. Would like to combine a `--files-from=` with `--missing-on-src`, but no errors are reported if some of the files are not present.

**Comment by haruki860** [comment]
@ncw 
I am new to contributing to OSS, but if the issue is not already assigned, may I work on it?

**Comment by shaliniganni-2000** [comment]
Hi @ncw - if this hasn't been worked on, can I work on it. I am new to open source and would love to contribute.

**Comment by ncw** [comment]
@shaliniganni-2000 that would be great - thanks. Let me know if you need any help.

**Comment by fjih** [comment]
There's a bit more going on here than the changes suggested.

The codepath splits in March.makeListDir(). Depending on the flags --fast-list and --no-traverse, we may or may not execute the code Nick is referring to.
If we use --fast-list we would have to check the return of ListR against the --files-from list in a separate piece of code.
However, I'm not certain why --no-traverse is taken into consideration for the function to list the source directory. --no-traverse is supposed to be a destination specific flag. Perhaps a change to March.init() is needed where setting m.srcListDir shouldn't be affected by --no-traverse?

I think what is missing is that second part of the code I mentioned; the other codepath needs to check the full list of files against the --files-from list. It looks like this would be done in list.filterAndSortDir().

In short, the code suggest by Nick handles: given an expected list of files, check that each one exists
What's missing is: given a list of files that exist, check there aren't any missing from an expected list of files

**Comment by fjih** [comment]
On deeper inspection, maybe the code was intending to always use the suggested code path with Nick's changes if --files-from is used. Everything seems to intend to take that path if --files-from is set but only if --no-traverse is also provided. I'm not really clear on why --no-traverse is important for this.

The suggested code path starts from the list of expected files in --files-from and gets them from the filesystem. This is definitely semantically akin to --no-traverse (we're not reading the entire directory but rather only the files we're interested in), but again, that flag is supposed to only apply to the destination. It seems like for a --files-from run we need to always take that code path for the source irrespective of --no-traverse.

It could be that the other uses of the enclosing functions Walk() and NewDirTree() of walk.go are protected from taking the suggest code path by the --no-traverse flag, but I think we're going to need to help from the contributors to know how to understand all this.

**Comment by emyasa** [comment]
Hi @ncw , I ran into this use-case myself and decided to give it a try.

To handle the case with `--no-traverse` flag, I followed your guide and made the adjustments in the `MakeListR` function.

As for the case without `--no-traverse` flag, which was how I used to copy files. I looked into `makeListDir` and noticed that it's source/destination agnostic. So I added an `isSource` parameter and a combined condition with the `fi.Opt.FilesFromStrict` option to prevent it from the `DirSorted` fallback, and instead use the same approach with `MakeListR` when the conditions are met (is source and has the strict flag):

https://github.com/emyasa/rclone/blob/feat/files-from-strict/fs/march/march.go#L123-L128

https://github.com/emyasa/rclone/blob/feat/files-from-strict/fs/walk/walk.go#L581-L587

I've been using rclone for a while now, but I'm very much new to the codebase. I made a pull request in case we are still looking to address this use case.
